app-marketplace-v2-spec/architecture.md

```diff
@@ -0,0 +1,146 @@
+# Architecture.
+
+  * [Cloud Apps.](#cloud-apps)
+    + [Authentication with Mattermost.](#authentication-with-mattermost)
+    + [Authentication with Upstream Applications.](#authentication-with-upstream-applications)
+    + [Cloud App to Mattermost.](#cloud-app-to-mattermost)
+    + [Mattermost to Cloud App.](#mattermost-to-cloud-app)
+    + [Functions](#functions)
+    + [Interactive Inputs.](#interactive-inputs)
+      - [Slash Commands](#slash-commands)
+      - [Widgets](#widgets)
+      - [Dialogs](#dialogs)
+    + [Change notifications.](#change-notifications)
+    + [Autocomplete lookups](#autocomplete-lookups)
+    + [Lifecycle - Install](#lifecycle---install)
+    + [Lifecycle Uninstall.](#lifecycle-uninstall)
+    + [Configuration.](#configuration)
+    + [mattermost-cloud-app.json](#mattermost-cloud-appjson)
+  * [mattermost-plugin-cloud-apps](#mattermost-plugin-cloud-apps)
+  * [mattermost-server](#mattermost-server)
+    + [Minimal Changes to the Core Code.](#minimal-changes-to-the-core-code)
+  * [Cloud App Marketplace](#cloud-app-marketplace)
+
+## Cloud Apps.
+
+### Authentication with Mattermost.
+1. OAuth2 will be used as a primary mechanism of authentication for the Mattermost REST API.
+2. Installing a **Cloud App** onto a Mattermost instance will provision sufficient credential to the app to invoke the REST API as the bot (priviledged? scoped? #TODO), and to verify/decode the JWT on the incoming requests from the Mattermost Server
+3. We should enable 3 use-cases:
+  a. Cloud App** using a "bot"-level token (to use APIs as the bot)
+  b. **Cloud App** using user-level OAuth2 tokens. Will require users' authoriation when first using the app.
```

Comment:
OAuth2 token will grant an app right to use user's data, act as a user, etc. Do we want to make it granular? For example "app may post only in the specific channels or DMs", "app gets access only to my public posts". Basically how detailed will be the scheme here?